blog/_src/posts/2016-11-02-beta-reduction-part-1.md

```diff
@@ -49,17 +49,17 @@ With the substitution function defined, we can write a semantics for evaluation:
+These rules mean that if you have a call in which you have a function applied to the argument, you substitute the argument into the function.  The next two rules say that if you can apply an evaluation rule to either the first or second subterm, you may do so.  Notice that both the second and third rules might apply to a single term.  These rules are nondeterministic and in these cases it is fine to use whichever rule you want (see below).
 
 # What is β-reduction?
 
-More generally, what is reduction?  Intuitively, a reduction system is a set of rules that determine how a term is stepped forwards in the computation.  β-reduction is Church's name for removing the λ from a function and substituting the argument for the parameter in the function body.  More formally, we can define β-reduction as follows:
+More generally, what is reduction?  Reduction is a model for computation that consists of a set of rules that determine how a term is stepped forwards.  β-reduction is reduction by function application.  When you β-reduce, you remove the λ from the function and substitute the argument for the function's parameter in its body.  More formally, we can define β-reduction as follows:
 
 ```
 (λx.e1) e2 = e1[ x := e2 ]
 ```
 
-Given that definition, lets look at a simple example.  I've written the examples in the λ-calculus, Racket, OCaml, and Haskell.  It's important to note that these languages have more restricted semantics than the original λ-calculus.  In Racket and OCaml, it is not possible to substitute with anything except for a value, meaning you need to evaluate the argument until it is a function before substituting.  This makes the evaluation reduce left to right before substituting.  Haskell, in contrast, no reduction occurs in arguments, meaning that we would omit the third rule.  
+Given that definition, lets look at a few examples.  I've written the examples in the λ-calculus, Racket[1], OCaml, and Haskell.  It's important to note that these languages have more restricted semantics than the original λ-calculus.  These restrictions are called reduction strategies.  In Racket and OCaml, it is not possible to substitute with anything except for a value, meaning you need to evaluate the argument until it is a function before substituting.  This makes the evaluation reduce left to right before substituting.  This restriction is called "call by value".  In Haskell, no reduction occurs in arguments, meaning that we would omit the third rule.  This could potentially require an expression to be evaluated multiple times.  In reality, Haskell caches the results of these evaluations so each expression is only evaluated once, but I'm going to ignore that here and give you the more theoretical version.  The theoretical version of Haskell's semantics is called "call by name" and the optimization is called "call by need".
```

Comment:
"theoretical version" is not quite right given that there have been theoretical studies of call-by-need which are quite interesting (two well-known small-step semantics are the ones of Ariola and Felleisen, and of Maraist, Odersky and Wadler; you could give a link to [The call-by-need lambda-calculus, revisited](http://www.ccs.neu.edu/racket/pubs/esop12-cf.pdf) from Chang and Felleisen which gives a good presentation of the historical approaches and a new solution). You could instead say that you give a "simple" or "naive" description of non-strict evluation, as opposed to the "optimized presentation" that is call-by-need. (Generally "presentation" or "description" would work better than "version" here).